frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java

```diff
@@ -44,23 +46,46 @@ public void addPlaceVisit(PlaceVisitModel placeVisit) throws PlaceVisitAlreadyEx
   }
 
   @Override
-  public void removePlaceVisit(String tripId, String placeId) throws PlaceVisitNotFoundException {
+  public void removePlaceVisit(String tripId, String placeUuid) throws PlaceVisitNotFoundException {
     Map<String, PlaceVisitModel> placesMap = placesByTripIdByPlaceId.get(tripId);
     if (placesMap == null) {
       throw new PlaceVisitNotFoundException(
-          "PlaceVisit with id" + placeId + " not found for trip " + tripId);
+          "PlaceVisit with id: '" + placeUuid + "' not found for trip '" + tripId + "'");
     }
 
-    if (placesMap.remove(placeId) == null) {
+    if (placesMap.remove(placeUuid) == null) {
       throw new PlaceVisitNotFoundException(
-          "PlaceVisit with id" + placeId + " not found for trip " + tripId);
+          "PlaceVisit with id: '" + placeUuid + "' not found for trip '" + tripId + "'");
     }
   }
 
   @Override
-  public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeId) {
+  public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeUuid) {
     return Optional.ofNullable(placesByTripIdByPlaceId.get(tripId))
-        .map(placesMap -> placesMap.get(placeId));
+        .map(placesMap -> placesMap.get(placeUuid));
+  }
+
+  @Override
+  public PlaceVisitModel updateUserMark(String tripId, String placeUuid,
+      PlaceVisitModel.UserMark newStatus) throws PlaceVisitNotFoundException {
+    AtomicReference<PlaceVisitModel> updatedPlaceVisit = new AtomicReference();
+
+    placesByTripIdByPlaceId.computeIfPresent(tripId, (tripKey, placesMap) -> {
+      PlaceVisitModel placeVisit = placesMap.get(placeUuid);
+      if (placeVisit != null) {
+        PlaceVisitModel newPlaceVisit = placeVisit.toBuilder().setUserMark(newStatus).build();
+        placesMap.put(placeUuid, newPlaceVisit);
+        updatedPlaceVisit.set(newPlaceVisit);
+      }
+      return placesMap;
+    });
+
+    if (updatedPlaceVisit.get() == null) {
+      throw new PlaceVisitNotFoundException(
+          "PlaceVisit with id: '" + placeUuid + "' not found for trip '" + tripId + "'");
```

Comment:
FYI, you can use `String.format` to make this more readable, e.g.:

```java
throw new PlaceVisitNotFoundException(String.format("PlaceVisit with id '%s' not found for trip '%s'"), placeUuid, tripId);
```

Or, since this exception is so specific to this condition, you could modify the exception to accept a trip id and place id directly and have the exception compute its own message.